rfcs/20200525-elasticity-and-fault-tolerance-in-multiworkermirroredstrategy.md

```diff
@@ -0,0 +1,439 @@
+
+```
+tensorflowrun --min_nodes Y --max_nodes Z \
+    --discovery disco_script.py --hostfile hostfile python train_script.py
+```
+
+Basic design:
+
+* run discovery script
+* Parse hostfile
+* Create inital per worker cluster_spec
+    * run training script with available hosts.
+* keep running discovery script
+    * if new hosts, start training script with updated cluster spec
+
+## Programming Flow
+
+* Create Training script as you would for tf.MultiWorkerMirroredStrategy
+* Create discovery script for your cluster
+* Use tensorflowrun to run your script [this can be a specific, or general cli]
+    * tensorflowrun —discovery_script discovery_script.sh —min_nodes X —max_nodes Y python training_script.py
+
+### Alternatives Considered
+
+* Similar to the above design, but only save the state before every run call (This is usually the start of an epoch)
+    * Before executing run call, Save state
+    * During runs, before OutOfRangeException
+        * If any nodes fail, Run membership process, restore all nodes to saved state, continue
+    * Frequent node failures, and longer epochs are prone to restarts and less progress will be made
+* Similar to the above design, but on membership changes, completely rebuild the whole session with the updated cluster spec. 
+    * This will incur any huge startup costs, which up to 3 mins is normal
+        * https://github.com/tensorflow/tensorflow/issues/25057
+* Add Checkpointing to Strategy
+    * On failure Auto restart from Checkpoint
+        * Restarting can either No session rebuild or Full session rebuild
+    * Documentation of proper usage of Saving state (should only be saved at start/end of epoch)
+
+### Performance Implications
+
+* Speed Implications
+    * As variables are being updated, there will be a backup process to enable recovery. This process will have performance implications. 
+    * Additional network calls will impact communication time
+    * Time taken to modify cluster size, initialize variables of new nodes will also impact performance
+* Memory implications
+    * CPU memory will increase as we save the variables in CPU memory.
+* Benchmarks comparing dedicated nodes vs nodes with planned failure to measure impact of elasticity
+* End to end tests will utilize the Multiprocess runner and run small training jobs with planned failures. These tests will be added to collective_all_reduce_strategy_test.py
+
+### Tutorials and Examples
+
+* The run script to specify min/max nodes will be a new cli
+    * tensorflowrun —min_nodes X —max_nodes Y —discovery_script discover.sh —hostfile hostfile python training_script.py
+* Alternative
+* `multiworker_strategy = tf.distribute.experimental.MultiWorkerMirroredStrategy(min_nodes=X, max_nodes=Y)`
+* A tutorial on creating discovery scripts
```

Comment:
TF ClusterResolver supports ClusterSpec update and propagation. Is there any other specific features we need to support using a discovery script, and if there is, could we extend the current TF ClusterResolver?

Reply:
Do you mean to bake in the discovery script into the TF Cluster resolver? That sounds good to me. 
By update and propagation, could you point me to where this is? I wasn't able to find this feature. Also do you mean update and propagate the cluster spec or the underlying grpc server?